bittide-instances/exe/post-vex-riscv-tcp-test/Main.hs

```diff
@@ -80,51 +99,78 @@ case_testGdbProgram = do
-        putStrLn " Done"
+        putStrLn "Waiting for Picocom to be ready..."
+        tryWithTimeout "Picocom handshake" 10_000_000 $
+          waitForLine picocomStdOutHandle "Terminal ready"
 
         putStrLn "Starting GDB..."
         withCreateProcess gdbProc $ \_ (fromJust -> gdbStdOut) _ _ -> do
-          -- Wait for GDB to program the FPGA. If successful, we should see
-          -- "going in echo mode" in the picocom output.
           hSetBuffering gdbStdOut LineBuffering
 
-          putStr "Waiting for GDB to program the FPGA..."
-          hFlush stdout
-          waitForLine picocomStdOutHandle "listening"
-          putStrLn " Done"
-
-        putStrLn "Starting TCP Spray..."
-        -- Test TCP echo server
-        withCreateProcess tcpSprayProc $ \_ (fromJust -> tcpStdOut) (fromJust -> tcpStdErr) _ -> do
-          (stdOutContents, stdErrContents) <- do
-            stdOutContents' <- hGetContents' tcpStdOut
-            stdErrContents' <- hGetContents' tcpStdErr
-            pure (stdOutContents', stdErrContents')
-          picocomOut <- hGetContents picocomStdOutHandle
-
-          case maybePicocomStdErr of
-            Nothing -> pure ()
-            Just h -> do
-              putStrLn "Picocom StdErr"
-              readRemainingChars h >>= putStrLn
-
-          putStrLn ""
-          putStrLn "Picocom stdout"
-          putStrLn $ readUntil "listening" picocomOut
-
-          putStrLn ""
-          putStrLn "tcpSpray StdOut"
-          putStrLn stdOutContents
-
-          if not (null stdErrContents)
-            then assertFailure $ "tcpspray failed with stderr:\n" <> stdErrContents
-            else pure ()
+          putStrLn "Waiting for \"Starting TCP Client\""
+          tryWithTimeout "Handshake softcore" 10_000_000 $
+            waitForLine picocomStdOutHandle "Starting TCP Client"
+
+          let numberOfClients = 1
+          putStrLn $ "Waiting for " <> show numberOfClients <> " clients to connect to TCP server."
+          clients <-
+            tryWithTimeout "Wait for clients" 10_000_000 $
+              waitForClients numberOfClients serverSock
+
+          putStrLn "Writing client data to files"
+          createDirectoryIfMissing True tcpDataDir
+          mapConcurrently_ runTcpTest clients
+          putStrLn "Closing client connections"
+          mapConcurrently_ (NS.closeSock . fst) clients
+          loggingSequence
+
+runTcpTest :: (NS.Socket, NS.SockAddr) -> Assertion
+runTcpTest (sock, sockAddr) = do
+  putStrLn $ "Testing TCP connection from " <> show sockAddr
+  start <- getPOSIXTime
+  bs <- SBS.toLazy_ $ SBS.reread (const $ NS.recv sock (1024 * 1024)) ()
+  stop <- getPOSIXTime
+  let
+    size = BS.length bs
+    diff = nominalDiffTimeToSeconds $ start - stop
+    speed = fromIntegral size / diff
+  putStrLn $
+    show sockAddr
+      <> " sent "
+      <> show size
+      <> " bytes in "
+      <> show diff
+      <> " seconds ("
+      <> show speed
+      <> " bytes/s)"
```

Comment:
Can you ease up on the precision of `diff` and `speed`? In [this run](https://github.com/bittide/bittide-hardware/actions/runs/11290288485/job/31406920217#step:9:138) the following line is printed:
`10.0.0.2:1234 sent 20294504 bytes in -28.312381858000 seconds (-716806.664369905236 bytes/s)`

Reply:
Not without introducing a new package, which seems a bit overkill tbh